applications/OptimizationApplication/python_scripts/algorithms/algorithm.py

```diff
@@ -0,0 +1,72 @@
+from abc import ABC
+from abc import abstractmethod
+
+import KratosMultiphysics as Kratos
+from KratosMultiphysics.python_solver import PythonSolver
+from KratosMultiphysics.OptimizationApplication.optimization_info import OptimizationInfo
+from KratosMultiphysics.OptimizationApplication.model_part_controllers.model_part_controller import ModelPartController
+from KratosMultiphysics.OptimizationApplication.utilities.control_transformation_technique import ControlTransformationTechnique
+from KratosMultiphysics.OptimizationApplication.utilities.response_function_implementor import ObjectiveResponseFunctionImplementor
+from KratosMultiphysics.OptimizationApplication.utilities.response_function_implementor import ConstraintResponseFunctionImplementor
+from KratosMultiphysics.OptimizationApplication.utilities.helper_utilities import CallOnAll
+
+class Algorithm(PythonSolver, ABC):
+    def __init__(self, model: Kratos.Model, parameters: Kratos.Parameters, optimization_info: OptimizationInfo):
+        PythonSolver.__init__(self, model, parameters)
+
+        self.optimization_info = optimization_info
+        self.optimization_info.SetBufferSize(self.GetMinimumBufferSize())
+        self.step = 0
+        self.optimization_info["step"] = self.step
+
+        self.__list_of_control_transformation_techniques = [self.optimization_info.GetOptimizationProcess(ControlTransformationTechnique, control_name) for control_name in parameters["control_names"].GetStringArray()]
+
+    def ImportModelPart(self):
+        CallOnAll(self.optimization_info.GetOptimizationProcesses(ModelPartController), ModelPartController.ImportModelPart)
+
+    def AdvanceInTime(self, _):
+        self.step += 1
+
+        # set the model part for the next iteration by applying update from the previous iteration
+        if self.step > 1:
+            CallOnAll(self.GetControlTransformationTechniques(), ControlTransformationTechnique.ApplyControlUpdate)
+
+        # now we march into next iteration
+        self.optimization_info.AdvanceSolutionStep()
+        self.optimization_info["step"] = self.step
+
+    def Initialize(self):
+        self.__list_of_objectives = [ObjectiveResponseFunctionImplementor(objective_settings, self.optimization_info) for objective_settings in self.settings["objectives"]]
+        self.__list_of_constraints = [ConstraintResponseFunctionImplementor(constraint_settings, self.optimization_info) for constraint_settings in self.settings["constraints"]]
```

Comment:
This should be moved to respective algorithms since, the standardization of each response may differ from algorithm to algorithm.

Reply:
Remove all the dependencies  of it as well.